signer.js

```diff
@@ -1,32 +1,96 @@
 'use strict'
 
-const cookieSignature = require('cookie-signature')
+const crypto = require('crypto')
 
-module.exports = function (secret) {
-  const secrets = Array.isArray(secret) ? secret : [secret]
-  const [signingKey] = secrets
+function SignerFactory (secrets, algorithm = 'sha256') {
+  if (!(this instanceof SignerFactory)) {
+    return new SignerFactory(secrets, algorithm)
+  }
 
-  return {
-    sign (value) {
-      return cookieSignature.sign(value, signingKey)
-    },
-    unsign (signedValue) {
-      let valid = false
-      let renew = false
-      let value = null
-
-      for (const key of secrets) {
-        const result = cookieSignature.unsign(signedValue, key)
-
-        if (result !== false) {
-          valid = true
-          renew = key !== signingKey
-          value = result
-          break
-        }
-      }
+  this.secrets = Array.isArray(secrets) ? secrets : [secrets]
+  this.signingKey = this.secrets[0]
+  this.algorithm = algorithm
 
-      return { valid, renew, value }
+  validateSecrets(this.secrets)
+  validateAlgorithm(this.algorithm)
+}
+
+function validateSecrets (secrets) {
+  for (const secret of secrets) {
+    if (typeof secret !== 'string') {
+      throw new TypeError('Secret key must be a string.')
     }
   }
 }
+
+function validateAlgorithm (algorithm) {
+  // validate that the algorithm is supported by the node runtime
+  try {
+    crypto.createHmac(algorithm, 'dummyHmac')
+  } catch (e) {
+    throw new TypeError(`Algorithm ${algorithm} not supported.`)
+  }
+}
+
+SignerFactory.prototype.sign = function (value, secret = this.signingKey, algorithm = this.algorithm) {
+  if (typeof value !== 'string') {
+    throw new TypeError('Cookie value must be provided as a string.')
+  }
+  return value + '.' + crypto
+    .createHmac(algorithm, secret)
+    .update(value)
+    .digest('base64')
+    // remove base64 padding (=) as it has special meaning in cookies
+    .replace(/=+$/, '')
+}
+
+SignerFactory.prototype.unsign = function (signedValue, secrets = this.secrets, algorithm = this.algorithm) {
+  if (typeof signedValue !== 'string') {
+    throw new TypeError('Signed cookie string must be provided.')
+  }
+  const value = signedValue.slice(0, signedValue.lastIndexOf('.'))
+  const actual = Buffer.from(signedValue)
+
+  for (const secret of secrets) {
+    const expected = Buffer.from(this.sign(value, secret, algorithm))
+    if (
+      expected.length === actual.length &&
+      crypto.timingSafeEqual(expected, actual)
+    ) {
+      return {
+        valid: true,
+        renew: secret !== secrets[0],
+        value
+      }
+    }
+  }
+
+  return {
+    valid: false,
+    renew: false,
+    value: null
+  }
+}
+
+// create a signer-instance, with dummy secret
+const signer = new SignerFactory(['dummy'])
```

Comment:
If I'm reading this code correctly, `dummy` would be used for `sign` if no secret is passed in the module-level `sign` function. This is unsafe, we should not have a default secret like this.

Reply:
@mcollina 

That is not correct. For the exposed shorthands sign and unsign, they will call validateSecrets before calling the underlying sign and unsign functions of the Signer. So there is now way, that the dummy key will be used.

Reply:
It's still pretty bad to see and look, and it's likely to come bite us in the future. Please fix.

Reply:
done